WordPress/gutenberg
# Borders: Add new BorderControl component

Related: 
- https://github.com/WordPress/gutenberg/issues/35602
- https://github.com/WordPress/gutenberg/pull/38876

## Description

This PR adds a new `BorderControl` component aimed at improving the UX around configuring borders.

- `BorderControl`
    - This allows the selection of a border color, style, and width.
    - Works off an object as its value e.g. `{ color: '#000000', style: 'solid', width: '0.5em' }`
    - Can optionally display a slider for alternative means of setting border width.
    - Border style controls can be disabled
    - The color palettes used are configurable
    - The width of the input field is customizable

#### Note
_The PR began life introducing two new components, `BorderControl` and `BorderBoxControl`. They have now been separated into separate PRs. Some of the implementation decisions in this were driven by the component's bigger picture use as part of the `BorderBoxControl`_

## How has this been tested?
Via unit tests;

1. Run `npm run test-unit packages/components/src/border-control/test/index.js`

Manually:

1. Checkout this PR branch
2. Run `npm run storybook:dev`
3. [Check out the `BorderControl`](http://localhost:50240/?path=/story/components-experimental-bordercontrol--default)

## Screenshots

<img src="https://user-images.githubusercontent.com/60436221/151919644-e661afd2-b777-45b1-ba5f-733e13161503.gif" width="500"/>

## Types of changes
New feature.

## Checklist:
